gemato/openpgp.py

```diff
@@ -589,18 +589,25 @@ def refresh_keys_wkd(self):
             raise_on_error=OpenPGPKeyRefreshError)
 
         # we need to explicitly ensure all keys were fetched
+        # However, a valid key *may* appear multiple times
+        # and other keys may be present (and need to be removed)
+        unexpected_keys = set()
         for line in out.splitlines():
             if line.startswith(b'[GNUPG:] IMPORT_OK'):
                 fpr = line.split(b' ')[3].decode('ASCII')
                 logging.debug(
                     f'refresh_keys_wkd(): import successful for key: {fpr}')
-                if fpr in keys:
-                    keys.remove(fpr)
-                else:
-                    # we need to delete unexpected keys
-                    exitst, out, err = self._spawn_gpg(
-                        [GNUPG, '--batch', '--delete-keys', fpr],
-                        raise_on_error=OpenPGPKeyRefreshError)
+                if fpr not in keys:
+                    unexpected_keys.add(fpr)
+
+        for fpr in unexpected_keys:
+            # we need to delete unexpected keys
+            logging.info(
+                f'refresh_keys_wkd(): got unexpected key, will remove: {fpr}')
+            exitst, out, err = self._spawn_gpg(
+                [GNUPG, '--batch', '--delete-keys', fpr],
```

Comment:
Hmm, can't you pass all keys here in one invocation?

Reply:
The prior code was deleting them one at a time, and I opted for as few changes as possible. 

If we did pass many keys at once, we'd need to ensure we didn't exceed arg max. Happy to include if you want that change in this PR as well, vs split out to another PR.

Reply:
implemented, with batches.

Reply:
I don't think it makes sense to add this level of complexity to account for removing 128+ keys which is unlikely enough to ever happen (especially giving WKD limits).

Reply:
So good to just drop the batch stuff entirely, and let it call single deletes?